File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.cc
#include <iostream>
#include <queue>
#include <cstdbool>

using namespace std;

queue<pair<int,int>> q;
int moveMap[101];
bool visited[101];

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int n, m;
	cin >> n >> m;
	for (int i = 0; i < n + m; i++) {
		int x, y;
		cin >> x >> y;
		moveMap[x] = y;
	}

	q.push({ 1,0 });
	visited[1] = true;

	while (!q.empty()) {
		int cur = q.front().first;
		int cnt = q.front().second;
		q.pop();
		if (cur == 100) {
			cout << cnt;
			break;
		}

		for (int i = 1; i <= 6; i++) {
			int nextCur = cur + i;
			if (nextCur > 100 || visited[nextCur])
				continue;
			if (moveMap[nextCur] != 0)
				nextCur = moveMap[nextCur];

			q.push({ nextCur,cnt + 1 });
			visited[nextCur] = true;
		}
	}
	return 0;
}

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
import sys
from collections import deque
input = sys.stdin.readline

# input
moveMap = {}
n, m = map(int, input().split())
for _ in range(n + m):
    x, y = map(int, input().split())
    moveMap[x] = y

visited = [False] * 101
queue = deque([(1, 0)])
visited[1] = True

# bfs
while queue:
    cur, cnt = queue.popleft()
    if cur == 100:
        print(cnt)
        break

    for i in range(1, 7):
        new = cur + i
        if new > 100 or visited[new]:
            continue
        if new in moveMap:
            new = moveMap[new]

        queue.append((new, cnt + 1))
        visited[new] = True
